mitiq/zne/scaling/layer_scaling.py

```diff
@@ -0,0 +1,78 @@
+# Copyright (C) 2023 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions for layer-wise unitary folding on supported circuits."""
+from typing import Callable, List
+import cirq
+from cirq import Moment, inverse
+
+from mitiq import QPROGRAM
+from mitiq.interface import noise_scaling_converter
+
+
+@noise_scaling_converter
+def layer_folding(
+    circuit: cirq.Circuit, layers_to_fold: List[int]
+) -> cirq.Circuit:
+    """Applies a variable amount of folding to select layers of a circuit.
+
+    Args:
+        circuit: The input cirq circuit.
+        layers_to_fold: A list with the index referring to the layer number,
+                        and the element filled by an integer represents the
+                        number of times the layer is folded.
+
```

Comment:
There is a small ambiguity about the length of `layers_to_fold`: do we count measurements as a layer?
Based on the code the answer seems to be :  yes, but the associated value should be 0 since in practice we cannot fold measurement layers.

Another option is to copy the workflow used in other folding functions where `check_foldable` and `pop_measurements` and `append_measurements` are called. In this case, `layers_to_fold` should be defined without considering the measurement layer (if present).

Whatever approach we use, we should probably say in the docstring what is the expected structure of `layers_to_fold` when measurements are present in the circuit.

Reply:
I think I like the second option you proposed here for the sake of consistency. I attempted to mimick the way in which it was done elsewhere within the folding function. Hopefully, that's what you had in mind. Thanks for the suggestion!